sql: don't cache uncommitted privileges

RichardJCai opened the pull request:
Release note (sql change): Previously, uncommitted privileges could be cached if the txn is rolled back. This is now fixed. This has not made it into GA and is a GA blocker. Example:

```
BEGIN;
GRANT SELECT ON crdb_internal.tables TO testuser;
SELECT has_table_privilege('testuser', 'crdb_internal.tables', 'SELECT'); --- this caches the privilege ---
ROLLBACK;
--- SELECT IS STILL CACHED UNTIL ANOTHER GRANT/REVOKE HAPPENS TO INVALIDATE THE CACHE---
```

Fixes: #89001

RichardJCai committed:
sql: don't cache uncommitted privileges

Release note (bug fix): Previously, uncommitted privileges could be cached if the txn is rolled back.
This is now fixed. This bug was only present in the 22.2 alpha and beta versions.
Example:

```
BEGIN;
GRANT SELECT ON crdb_internal.tables TO testuser;
SELECT has_table_privilege('testuser', 'crdb_internal.tables', 'SELECT'); --- this caches the privilege ---
ROLLBACK;
--- SELECT IS STILL CACHED UNTIL ANOTHER GRANT/REVOKE HAPPENS TO INVALIDATE THE CACHE---
```

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/89235)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
thanks for getting this!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-:-NDUgOQ3-7W1Rn1ihgF9:bpfc5oe)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @RichardJCai)

___
*[`-- commits` line 4 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUexAY4KLS00ddM42C):*
nit: should be "bug fix" not sql change
___
*[`-- commits` line 5 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUezkD2bvqFfK1goK-):*
nit: customers don't know what GA blocker means for us, so can this be rephrased as "This bug is only present in v22.2 alpha and beta releases"
___
*[`pkg/sql/grant_revoke_system.go` line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUfXs7ADPgbad0Th8g:-NDUfXs8BT5t7i_Q7Gr3:b7posc0) ([raw file](https://github.com/cockroachdb/cockroach/blob/864f4079c855d0fb37a0c32e7b38a06743b1453c/pkg/sql/grant_revoke_system.go#L256)):*
> ```Go
> 		cache.Lock()
> 		defer cache.Unlock()
> 		_, desc, err := p.Descriptors().GetImmutableTableByName(ctx, p.Txn(),
> ```

nit: loading the descriptor can be done outside of the area where we are holding the lock. making this change should reduce some contention

i think if the descriptor is uncommitted, we should also skip the logic to clear the cache. see how authorization.go works 
___
*[`pkg/sql/grant_revoke_system.go` line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUg9DS-WBOH0F2jtow:-NDUg9DW3f1k2dhPLhaY:b-sfyl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/864f4079c855d0fb37a0c32e7b38a06743b1453c/pkg/sql/grant_revoke_system.go#L279)):*
> ```Go
> 	}
> 
> 	val, err := cache.LoadValueOutsideOfCache(ctx, privilegeObjectPath, func(loadCtx context.Context) (_ interface{}, retErr error) {
> ```

this `func` can be extracted as a variable or its own function somewhere. then we can call it directly if the descriptor is uncommitted, and short-circuit all the other caching logic
___
*[`pkg/sql/grant_revoke_system.go` line 369 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUgJaZ9ar3VnS-Ht7f:-NDUgJaZ9ar3VnS-Ht7g:b9s04sx) ([raw file](https://github.com/cockroachdb/cockroach/blob/864f4079c855d0fb37a0c32e7b38a06743b1453c/pkg/sql/grant_revoke_system.go#L369)):*
> ```Go
> 	// Only write back to the cache if the table version is
> 	// committed.
> 	if isCommitted {
> ```

nit: wouldn't need this if we short-circuit


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @rafiss)

___
*[`pkg/sql/grant_revoke_system.go` line 256 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUfXs7ADPgbad0Th8g:-NDUkGuz8XNVhZ6E1MrU:bb6un1j) ([raw file](https://github.com/cockroachdb/cockroach/blob/864f4079c855d0fb37a0c32e7b38a06743b1453c/pkg/sql/grant_revoke_system.go#L256)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: loading the descriptor can be done outside of the area where we are holding the lock. making this change should reduce some contention

i think if the descriptor is uncommitted, we should also skip the logic to clear the cache. see how authorization.go works 
</blockquote></details>

Done
___
*[`pkg/sql/grant_revoke_system.go` line 279 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUg9DS-WBOH0F2jtow:-NDUkHb79GdKEMZSLX2u:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/864f4079c855d0fb37a0c32e7b38a06743b1453c/pkg/sql/grant_revoke_system.go#L279)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this `func` can be extracted as a variable or its own function somewhere. then we can call it directly if the descriptor is uncommitted, and short-circuit all the other caching logic
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

ajwerner commented on pkg/sql/grant_revoke_system.go:
```diff
@@ -247,25 +247,28 @@ func (p *planner) SynthesizePrivilegeDescriptor(
+	if err != nil {
+		return nil, err
+	}
+	if !desc.IsUncommittedVersion() {
```
Can you help me understand this code? If it is **not** an uncommitted version, then we'll go to this method which is going to synchronize with concurrent cache updates but we'll not write that to the cache. Then if it **is** an uncommitted version (which means that there are changes, right?), then we will check the cache and maybe write it back?

ajwerner commented on pkg/sql/grant_revoke_system.go:
```diff
@@ -274,7 +277,24 @@ func (p *planner) SynthesizePrivilegeDescriptor(
+	privilegeObjectPath string,
+	privilegeObjectType privilege.ObjectType,
+) (*catpb.PrivilegeDescriptor, error) {
+	val, err := p.ExecCfg().SyntheticPrivilegeCache.LoadValueOutsideOfCache(ctx, privilegeObjectPath, func(loadCtx context.Context) (_ interface{}, retErr error) {
```
I still feel uncomfortable that we're using the user's transaction at all to read from the cache. I feel much worse about the idea that we're going to use the cache for its singleflight unconditionally even when we're trying to avoid using the cache.

ajwerner replied:
If we're avoiding the cache we *must* use the txn for the session (it might have performed writes). If we are avoiding the cache we definitely should not block others who are using the cache.

RichardJCai replied:
I've updated it to use the txn for the session

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)

___
*[`pkg/sql/grant_revoke_system.go` line 297 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUYNmcUHXMz8N98kg3-r2-297:-NDZ8psKE8d7b4w4EP98:b7izzyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f5d5085cf5e72e01f37fef0a1ca91a69a20927/pkg/sql/grant_revoke_system.go#L297)):*
<details><summary><i>Previously, RichardJCai (Richard Cai) wrote…</i></summary><blockquote>

I've updated it to use the txn for the session 
</blockquote></details>

as an aside, i just realized there was also a bug in the single-flight logic before. the `requestKey` used for the singleflight should have included the table version in it (similar to how sessioninit/cache.go and the cache in authorization.go do it)

but now i'm not following why we removed the usage of singleflight entirely here
___
*[`pkg/sql/grant_revoke_system.go` line 277 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDZ7cOgCASJl-RS-q6e:-NDZ7cOgCASJl-RS-q6f:bimy3p0) ([raw file](https://github.com/cockroachdb/cockroach/blob/9794be4c70f2591c36a3b1213ed3f2dd7ef0e481/pkg/sql/grant_revoke_system.go#L277)):*
> ```Go
> 	}
> 
> 	val, err := cache.LoadValueOutsideOfCache(ctx, privilegeObjectPath, func(loadCtx context.Context) (_ interface{}, retErr error) {
> ```

i think the old `LoadValueOutsideOfCache` now has no non-test usages
___
*[`pkg/sql/grant_revoke_system.go` line 309 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDZ7Cqc-j66c0HWBIYz:-NDZ7Cqe9uNzK3xgWJYS:b-iz199p) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef0169e9cc1dd8896fe014520a7c62a863b490f/pkg/sql/grant_revoke_system.go#L309)):*
> ```Go
> 	if err != nil {
> 		return nil, err
> 	}
> ```

this used to have a 

```
defer func() { retErr = errors.CombineErrors(retErr, it.Close()) }()
````

and we still need that


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)

___
*[`pkg/sql/grant_revoke_system.go` line 297 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUYNmcUHXMz8N98kg3-r2-297:-NDZ9U6qADX3bD9M5HJj:b3aeg4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f5d5085cf5e72e01f37fef0a1ca91a69a20927/pkg/sql/grant_revoke_system.go#L297)):*
> I still feel uncomfortable that we're using the user's transaction at all to read from the cache

Do you share that concern for the cache in authorization.go? could you say more what the concern is?

> I feel much worse about the idea that we're going to use the cache for its singleflight unconditionally even when we're trying to avoid using the cache.

definitely. if the descriptor is uncommitted, we shouldn't be using singleflight. (and to reference authorization.go again, that one is doing it correctly)


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)

___
*[`pkg/sql/grant_revoke_system.go` line 297 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDUYNmcUHXMz8N98kg3-r2-297:-NDZB-Nl-e1tCvWPBKnI:b-oqm5im) ([raw file](https://github.com/cockroachdb/cockroach/blob/32f5d5085cf5e72e01f37fef0a1ca91a69a20927/pkg/sql/grant_revoke_system.go#L297)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

> I still feel uncomfortable that we're using the user's transaction at all to read from the cache

Do you share that concern for the cache in authorization.go? could you say more what the concern is?

> I feel much worse about the idea that we're going to use the cache for its singleflight unconditionally even when we're trying to avoid using the cache.

definitely. if the descriptor is uncommitted, we shouldn't be using singleflight. (and to reference authorization.go again, that one is doing it correctly)
</blockquote></details>

For the first point, it's out of fear that we've messed something up regarding the descriptor versions. I'd like to see the cache component fully separated from the session in terms of dependencies to the extent possible. It's not as big of a deal. I guess that on some level I wonder whether we could extract these caches to a package other than sql. That's where my head was when I wrote that. 


<!-- Sent from Reviewable.io -->

RichardJCai reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)

___
*[`pkg/sql/grant_revoke_system.go` line 277 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDZ7cOgCASJl-RS-q6e:-ND_2P7QC-OkTJtSlZip:b5nwtu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/9794be4c70f2591c36a3b1213ed3f2dd7ef0e481/pkg/sql/grant_revoke_system.go#L277)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think the old `LoadValueOutsideOfCache` now has no non-test usages
</blockquote></details>

Removed completely for now
___
*[`pkg/sql/grant_revoke_system.go` line 309 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDZ7Cqc-j66c0HWBIYz:-ND_2hI-5vYuRSHSHROm:bzdbw89) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ef0169e9cc1dd8896fe014520a7c62a863b490f/pkg/sql/grant_revoke_system.go#L309)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this used to have a 

```
defer func() { retErr = errors.CombineErrors(retErr, it.Close()) }()
````

and we still need that
</blockquote></details>

True, done


<!-- Sent from Reviewable.io -->

RichardJCai commented:
>but now i'm not following why we removed the usage of singleflight entirely here

Removed for now for simplicity and being able to reason about correctness. I also don't think the use of singleflight is as important here as it is for authentication either. These privilege checks tend to be part of more expensive operations anyways.

ajwerner reviewed (commented):
I don't think I buy that. Don't we access this whenever we look at virtual tables? That's almost always the first thing that connections to. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-:-ND_5Nnx-U1Z3nZyz52K:bbxont5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)



<!-- Sent from Reviewable.io -->

RichardJCai commented:
> I don't think I buy that. Don't we access this whenever we look at virtual tables? That's almost always the first thing that connections to.
> 
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-:-ND_5Nnx-U1Z3nZyz52K:bbxont5)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)

Mhm yeah the vtable case is not great cause we don't short circuit it. How common would we be concurrently resolving the same descriptor. Singleflight only helps when we have multiple concurrent queries at the same time before we can even cache anything.

ajwerner commented:
I can absolutely imagine a flurry of new connections which read the vtable. It's the same situation we saw with the password hashes. I'd say that it is important.

rafiss commented:
+1 to keeping singleflight (especially since we already have that pattern working in the role membership cache). the case @ajwerner describes sounds common to me.

RichardJCai reviewed (commented):
Added back singleflight, I believe this is okay now, we use `nil` txn in the singleflight and the session txn for when we avoid the cache due to uncommitted table desc (in case writes happened)

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-:-NDdTkhuET0TdQf-puxO:bau7xac)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)



<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @RichardJCai)

___
*[`pkg/sql/grant_revoke_system.go` line 271 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDddHvX1kusQhFqfPMr:-NDddHvX1kusQhFqfPMs:b-772p5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b231dbfc5a46784c5075690ae86a260a381e6eae/pkg/sql/grant_revoke_system.go#L271)):*
> ```Go
> 			if ok {
> 				privilegeDescriptor := val.(catpb.PrivilegeDescriptor)
> 				return true, &privilegeDescriptor, nil
> ```

are we happy allocating on this path every time? 


<!-- Sent from Reviewable.io -->

RichardJCai commented:
> > _[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235)_ status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss and @RichardJCai)
> 
> _[`pkg/sql/grant_revoke_system.go` line 271 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDddHvX1kusQhFqfPMr:-NDddHvX1kusQhFqfPMs:b-772p5k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b231dbfc5a46784c5075690ae86a260a381e6eae/pkg/sql/grant_revoke_system.go#L271)):_
> 
> > ```go
> > 			if ok {
> > 				privilegeDescriptor := val.(catpb.PrivilegeDescriptor)
> > 				return true, &privilegeDescriptor, nil
> > ```
> 
> are we happy allocating on this path every time?

Removed alloc

rafiss commented:
I'm still not fully on board with the singleflight using a nil txn. Don't we need it to use one of the txns doing the privilege check, in order to be sure that it doesn't read data from a different version of the system.privileges table? This was the reasoning used for the role membership cache.

I do agree that singleflight should use a different context.Context though.

ajwerner commented:
> I'm still not fully on board with the singleflight using a nil txn. Don't we need it to use one of the txns doing the privilege check, in order to be sure that it doesn't read data from a different version of the system.privileges table? This was the reasoning used for the role membership cache.

As long as the read timestamp of the cache reading transaction is in the validity interval of the lease, there's not a problem. On some level, the best thing we could do is to set a fixed timestamp on that transaction. That would avoid any uncertainty problems -- the only thing which can cause a timestamp for a reading transaction to have to change.

RichardJCai commented:
> > I'm still not fully on board with the singleflight using a nil txn. Don't we need it to use one of the txns doing the privilege check, in order to be sure that it doesn't read data from a different version of the system.privileges table? This was the reasoning used for the role membership cache.
> 
> As long as the read timestamp of the cache reading transaction is in the validity interval of the lease, there's not a problem. On some level, the best thing we could do is to set a fixed timestamp on that transaction. That would avoid any uncertainty problems -- the only thing which can cause a timestamp for a reading transaction to have to change.

If we set the fixed timestamp, how do we guarantee that the timestamp we're setting to is in the validity of interval of the lease?

ajwerner commented:
> If we set the fixed timestamp, how do we guarantee that the timestamp we're setting to is in the validity of interval of the lease?

By construction, the transaction's read timestamp is in the validity interval.

RichardJCai commented:
> > If we set the fixed timestamp, how do we guarantee that the timestamp we're setting to is in the validity of interval of the lease?
> 
> By construction, the transaction's read timestamp is in the validity interval.

Alright so we can go back to using any `txn` that reads a committed `system.privileges` desc right?

ajwerner commented:
It's fine. I'll accept it. The modest downside is you pick up dependencies and potentially run into an uncertain write.

RichardJCai commented:
> It's fine. I'll accept it. The modest downside is you pick up dependencies and potentially run into an uncertain write.

What do you mean specifically by dependencies - that it's attached to a certain session?

Also what would be the workaround, create a new txn and set the fixed timestamp to another txns read timestamp?

ajwerner commented:
I mean the reads performed by that transaction will be recorded as reads by whatever transaction ended up doing the read. The caching semantics mean that no transaction needs to have read that data itself. When the transaction goes to commit, if it has to do a read refresh, will have to refresh that data.

RichardJCai commented:
> It's fine. I'll accept it. The modest downside is you pick up dependencies and potentially run into an uncertain write.

Alright updated to do this.

rafiss reviewed (approved):
just had small nits, lgtm!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-:-NDjKA9cDIK5dz8CqAsR:b-4tj9um)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @rafiss, and @RichardJCai)

___
*[`pkg/sql/grant_revoke_system.go` line 301 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDjJpBW3wMxki9rGtUw:-NDjJpBY4Tfnb8WvO85Q:b-r3q5z) ([raw file](https://github.com/cockroachdb/cockroach/blob/069a0949435df92e060f873106b381c05b253f40/pkg/sql/grant_revoke_system.go#L301)):*
> ```Go
> func (p *planner) synthesizePrivilegeDescriptorFromSystemPrivilegesTable(
> 	ctx context.Context,
> 	txn *kv.Txn,
> ```

nit: the `txn` here is always `p.Txn()` now, so it can be removed
___
*[`pkg/sql/grant_revoke_system.go` line 311 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/89235#-NDjJs-q0lJhwEYlSKu3:-NDjJs-q0lJhwEYlSKu4:bpfxgra) ([raw file](https://github.com/cockroachdb/cockroach/blob/069a0949435df92e060f873106b381c05b253f40/pkg/sql/grant_revoke_system.go#L311)):*
> ```Go
> 		privilegeObjectPath)
> 
> 	err := p.ExecCfg().InternalExecutorFactory.RunWithoutTxn(ctx, func(ctx context.Context, ie sqlutil.InternalExecutor) error {
> ```

nit: similar to above, we're always using `p.Txn` now, so we don't need the `RunWithoutTxn` helper


<!-- Sent from Reviewable.io -->

RichardJCai commented:
Thanks for reviewing!

bors r=rafiss
